hoomd/MeshDefinition.h

```diff
@@ -0,0 +1,111 @@
+/*! \file MeshDefinition.h
+    \brief Defines the MeshDefinition class
+ */
+
+#ifdef __HIPCC__
+#error This header cannot be compiled by nvcc
+#endif
+
+#include "BondedGroupData.h"
+#include "IntegratorData.h"
+#include "MeshGroupData.h"
+#include "SystemDefinition.h"
+
+#include <memory>
+#include <pybind11/pybind11.h>
+
+#ifndef __MESH_DEFINITION_H__
+#define __MESH_DEFINITION_H__
+
+namespace hoomd
+    {
+#ifdef ENABLE_MPI
+//! Forward declaration of Communicator
+class Communicator;
+#endif
+
+//! Mesh class that contains all infrastructure necessary to combine a set of particles into a mesh
+//! triangulation
+/*! MeshDefinition is a container class to define a mesh tringulation comprised of the
+    particles within the simulation system. The vertices of the mesh are given by the
+    particle positions. The information which vertex tuples are directly bonded (i.e. share an
+    edge) and which vertex triplets consitute triangles is stored as two shared pointer for
+    quick and easy access from within C++ and python without worrying about data management.
+    The pointers also encode simplex labels such that the we can detect easily which bonds make
+    up a triangle and which triangles share a common edge bond.
+
+
+    <b>Background and intended usage</b>
+
+    The data structure stored in MeshDefinition is all the edge bonds and triangles within
+    a particle mesh. It is used to combine a set of particles into a bonded mesh. In doing so
+    we can asign different potentials (as bond or surface potentials) to the mesh data structure.
+    The class has to access system information such as the particle positions to define the mesh
+    vertices.
+
+    As any data structure class in MeshDefinition can potentially reference any
+   other, other classes can simply use the mesh data by giving the shared pointer to the referenced
+   class to the constructor of the onem that needs to refer to it. Note that using this setup, there
+   can be no circular references. This is a \b good \b thing ^TM, as it promotes good separation
+   and isolation of the various classes responsibilities.
+
+    <b>Initializing</b>
+
+    A default constructed MeshDefinition is full of NULL shared pointers. Afterwards the user has to
+    specify the triangulation information via a Triangle snapshot (see MeshGroupData.h). The
+    corresponding edge information is calculated automatically. MeshDefinition allows for only one
+    mesh type to prevent ambiguity how to set triangle and edge bond types.
+
+
+*/
+class PYBIND11_EXPORT MeshDefinition
+    {
+    public:
+    //! Constructs a NULL MeshDefinition
+    MeshDefinition();
+    //! Constructs a MeshDefinition with a simply initialized ParticleData
+    MeshDefinition(std::shared_ptr<SystemDefinition> sysdef);
+
+    //! Access the mesh triangle data defined for the simulation
+    std::shared_ptr<MeshTriangleData> getMeshTriangleData()
+        {
+        return m_meshtriangle_data;
+        }
```

Comment:
This will be used mainly by the potentials which can be applied to the meshes later. The bond potential only needs mesh bonds. Other potentials like Helfrich bending potential or area conservation also have to get the mesh triangle data . I am working on those potentials in tandem with this PR.

Reply:
Makes sense. :+1: